src/postoffice.cc

```diff
@@ -25,78 +25,104 @@ Postoffice::Postoffice() {
-  for (int i = 0; i < num_workers_; ++i) {
-    int id = WorkerRankToID(i);
-    for (int g : {id, kWorkerGroup, kWorkerGroup + kServerGroup,
-            kWorkerGroup + kScheduler,
-            kWorkerGroup + kServerGroup + kScheduler}) {
-      node_ids_[g].push_back(id);
+    // init node info.
+    for (int i = 0; i < num_workers_; ++i) {
+      int id = WorkerRankToID(i);
+      for (int g : {id, kWorkerGroup, kWorkerGroup + kServerGroup,
+                    kWorkerGroup + kScheduler,
+                    kWorkerGroup + kServerGroup + kScheduler}) {
+        node_ids_[g].push_back(id);
+      }
     }
-  }
 
-  for (int i = 0; i < num_servers_; ++i) {
-    int id = ServerRankToID(i);
-    for (int g : {id, kServerGroup, kWorkerGroup + kServerGroup,
-            kServerGroup + kScheduler,
-            kWorkerGroup + kServerGroup + kScheduler}) {
-      node_ids_[g].push_back(id);
+    for (int i = 0; i < num_servers_; ++i) {
+      int id = ServerRankToID(i);
+      for (int g : {id, kServerGroup, kWorkerGroup + kServerGroup,
+                    kServerGroup + kScheduler,
+                    kWorkerGroup + kServerGroup + kScheduler}) {
+        node_ids_[g].push_back(id);
+      }
     }
-  }
 
-  for (int g : {kScheduler, kScheduler + kServerGroup + kWorkerGroup,
-          kScheduler + kWorkerGroup, kScheduler + kServerGroup}) {
-    node_ids_[g].push_back(kScheduler);
+    for (int g : {kScheduler, kScheduler + kServerGroup + kWorkerGroup,
+                  kScheduler + kWorkerGroup, kScheduler + kServerGroup}) {
+      node_ids_[g].push_back(kScheduler);
+    }
+    init_stage++;
   }
+  start_mu_.unlock();
 
   // start van
-  van_->Start();
-
-  // record start time
-  start_time_ = time(NULL);
-
-  // do a barrier here
-  if (do_barrier) Barrier(kWorkerGroup + kServerGroup + kScheduler);
+  van_->Start(customer_id);
+
+  start_mu_.lock();
+  if (init_stage == 1) {
+    // record start time
+    start_time_ = time(NULL);
+    // do a barrier here
+    init_stage++;
+  }
+  start_mu_.unlock();
+  if (do_barrier) Barrier(customer_id, kWorkerGroup + kServerGroup + kScheduler);
 }
 
-void Postoffice::Finalize(const bool do_barrier) {
-  if (do_barrier) Barrier(kWorkerGroup + kServerGroup + kScheduler);
-  van_->Stop();
-  if (exit_callback_) exit_callback_();
+void Postoffice::Finalize(const int customer_id, const bool do_barrier) {
+  if (do_barrier) Barrier(customer_id, kWorkerGroup + kServerGroup + kScheduler);
+  if (customer_id == 0) {
+    van_->Stop();
+    if (exit_callback_) exit_callback_();
+  }
 }
 
 
 void Postoffice::AddCustomer(Customer* customer) {
   std::lock_guard<std::mutex> lk(mu_);
-  int id = CHECK_NOTNULL(customer)->id();
-  CHECK_EQ(customers_.count(id), (size_t)0) << "id " << id << " already exists";
-  customers_[id] = customer;
+  int app_id = CHECK_NOTNULL(customer)->app_id();
+  if (customers_.find(app_id) == customers_.end()) {
+    customers_[app_id] = *(new std::unordered_map<int, Customer*>());
```

Comment:
memory leak

Reply:
how? please forgive that I have nearly forgotten everything of c++, but this should be alive as long as the whole process

Reply:
`customers_[...] = ` will call copy constructor of the right one, i.e., `*(new ...)`. Then you just lost the pointer and have no chance to free the memory.

Reply:
if I understand correctly, it is because of the [] operator of unordered_map: http://www.cplusplus.com/reference/unordered_map/unordered_map/operator[]/? (`If k does not match the key of any element in the container, the function inserts a new element with that key and returns a reference to its mapped value. Notice that this always increases the container size by one, even if no mapped value is assigned to the element (the element is constructed using its default constructor).`)

https://github.com/dmlc/ps-lite/pull/121/commits/56c90af29572b63d4f200627de42c1abc4c77dfb should fix it